client/state/sharing/publicize/actions.js

```diff
@@ -42,21 +42,33 @@ export function sharePost( siteId, postId, skippedConnections, message ) {
 			message,
 		} );
 
-		return new Promise( ( resolve ) => {
-			wpcom
-				.undocumented()
-				.publicizePost( siteId, postId, message, skippedConnections, ( error, data ) => {
-					// Note: successes are recorded in data.results, errors are recorded in data.errors. There could be
-					// several errors and several successes.
-					if ( error || ! data.results ) {
-						dispatch( { type: PUBLICIZE_SHARE_FAILURE, siteId, postId, error } );
-					} else {
-						dispatch( { type: PUBLICIZE_SHARE_SUCCESS, siteId, postId } );
-					}
+		const body = {
+			skipped_connections: [],
+			message,
+		};
 
-					resolve();
-				} );
-		} );
+		if ( skippedConnections && skippedConnections.length > 0 ) {
+			body.skipped_connections = skippedConnections;
+		}
+
+		wpcom.req
+			.post( {
+				path: `/sites/${ siteId }/posts/${ postId }/publicize`,
+				body,
+				apiNamespace: 'wpcom/v2',
+			} )
+			// Note: successes are recorded in data.results, errors are recorded in data.errors. There could be
+			// several errors and several successes.
+			.then( ( data ) => {
+				if ( ! data.results ) {
+					dispatch( { type: PUBLICIZE_SHARE_FAILURE, siteId, postId, error: true } );
+				} else {
+					dispatch( { type: PUBLICIZE_SHARE_SUCCESS, siteId, postId } );
+				}
+			} )
+			.catch( ( error ) => {
+				dispatch( { type: PUBLICIZE_SHARE_FAILURE, siteId, postId, error } );
+			} );
```

Comment:
The error handling here (both before and after the change) looks weird and I suspect it never worked.

First, the response object contains two arrays, `data.results` and `data.errors`. They are always present, even if empty. The `! data.results` condition is always `false`. The `/publicize` action publishes the post to a list of connections, e.g., social network accounts, and some can succeed and some can fail.

Second, the `error` field in the `PUBLICIZE_SHARE_FAILURE` action is expected to be a boolean. The `sharePostFailure` selector that reads the state written by this action does a precise `return state === true` check. But the error thrown by the promise can be a network or authentication error, and then the `error === true` evaluates to `false` even if there was error?

Reply:
Good point. I admit, my strategy was to migrate the logic 1:1, however it indeed looks like it's not going to work (neither before, nor after), so why not as well fix it.

In 85bec19de7df627dfdf4cad1b17a742eabc17467 I've updated the code to check for the `length` of the `results` array, instead of the array itself. This is still not ideal (due to the nature of the logic where we can have multiple errors and/or multiple successes), but it makes more sense IMHO.

Furthermore, in c8563c96b4a45742c7ae766d6fe9ac88143ad500 I've updated the error to be set as a boolean flag since it's consumed as one in the sole selector that uses that state value.

Reply:
So, if we manage to share the post to at least one account, then it's a success, and a failure is declared only when _all_ accounts fail? That's fair, most users probably have just one social account.

There's one more pre-existing bug in the `blocks/post-share` component: the `sharePostFailure` value is connected as a `failed` prop, but the component expects `failure`. The naming needs to be fixed, otherwise the error notice is never displayed.

> I've updated the error to be set as a boolean flag

That's good, but now when the `PUBLICIZE_SHARE_FAILURE` action always has an `error: true` field, the field no longer needs to be there.

Reply:
> There's one more pre-existing bug in the blocks/post-share component: the sharePostFailure value is connected as a failed prop, but the component expects failure. The naming needs to be fixed, otherwise the error notice is never displayed.

Fixed in https://github.com/Automattic/wp-calypso/pull/57644/commits/9685d6654a6215924dcb624a3a554b98fa66ee6b. I laughed a little bit at the error message here FWIW:

![](https://cldup.com/0e9mKVCqgg.png)

Please don't be mad! 😅

> That's good, but now when the PUBLICIZE_SHARE_FAILURE action always has an error: true field, the field no longer needs to be there.

Good point, I've cleaned that up in https://github.com/Automattic/wp-calypso/pull/57644/commits/2102d4daf2db691197028811ffd0e9d6d219774b

Reply:
> I laughed a little bit at the error message here FWIW

Oh yes, that's just one of several clues that @artpi is the author of this feature 🙂